### PR TITLE
Added node for Watson Conversation (Experimental)

### PR DESCRIPTION
**Pull request by ylecleach**

### CODE REVIEW

File: services/conversation/v1-exp.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    msg.result = body;
+    msg.payload = 'see msg.result';
+    node.send(msg); 
+    node.status({});
+  }
+
+  function execute(params, node, msg) {
+    node.status({fill:'blue', shape:'dot' , text:'Calling Conversation service ...'});
+    params.workspace_id=node.workspaceid;
```
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l101)

File: services/conversation/v1-exp.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+  
+  service = cfenv.getAppEnv().getServiceCreds(/conversation/i);
+
+  if (service) {
+    sUsername = service.username;
+    sPassword = service.password;
+  }
+
+  RED.httpAdmin.get('/watson-conversation/vcap', function (req, res) {
+    res.json(service ? {bound_service: true} : null);
```
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'bound_service' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l29)

File: services/conversation/v1-exp.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    
+    if (!userName || !passWord) {
+      node.status({fill:'red', shape:'ring', text:'missing credentials'});          
+      node.error('Missing Watson Conversation API service credentials', msg);
+      return false;
+    }
+    node.service = watson.conversation({
+      username: userName,
+      password: passWord,
+      version_date: '2016-05-19',
```
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'version_date' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l71)

File: services/conversation/v1-exp.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  function verifyInputs(node, msg, config) {
+    // workspaceid can be either configured in the node, 
+    // or sent into msg.params.workspace_id
+    if (!config.workspaceid || '' !== config.workspaceid) {
+      node.workspaceid = config.workspaceid;
+      return true;
+    } else if (!msg.params || '' !== msg.params.workspace_id) {
+      node.workspaceid = msg.params.workspace_id;
```
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l49)

File: services/conversation/v1-exp.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    return true;
+  }
+
+  function verifyInputs(node, msg, config) {
+    // workspaceid can be either configured in the node, 
+    // or sent into msg.params.workspace_id
+    if (!config.workspaceid || '' !== config.workspaceid) {
+      node.workspaceid = config.workspaceid;
+      return true;
+    } else if (!msg.params || '' !== msg.params.workspace_id) {
```
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l48)

File: services/conversation/v1-exp.js (hunk 1 above)
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l101)

File: services/conversation/v1-exp.js (hunk 2 above)
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'bound_service' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l29)

File: services/conversation/v1-exp.js (hunk 3 above)
Comment (chughts) [inline services/conversation/v1-exp.js]: Codacy Issue found: [Identifier 'version_date' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2775458781/issues/source?bid=3330694&fileBranchId=3363974#l71)

File: services/conversation/v1-exp.js (hunk 1 above)
Comment (chughts) [inline services/conversation/v1-exp.js:101]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l101)

File: services/conversation/v1-exp.js (hunk 2 above)
Comment (chughts) [inline services/conversation/v1-exp.js:29]: Codacy Issue found: [Identifier 'bound_service' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l29)

File: services/conversation/v1-exp.js (hunk 3 above)
Comment (chughts) [inline services/conversation/v1-exp.js:71]: Codacy Issue found: [Identifier 'version_date' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l71)

File: services/conversation/v1-exp.js (hunk 1 above)
Comment (chughts) [inline services/conversation/v1-exp.js:101]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l101)

File: services/conversation/v1-exp.js (hunk 2 above)
Comment (chughts) [inline services/conversation/v1-exp.js:29]: Codacy Issue found: [Identifier 'bound_service' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l29)

File: services/conversation/v1-exp.js (hunk 3 above)
Comment (chughts) [inline services/conversation/v1-exp.js:71]: Codacy Issue found: [Identifier 'version_date' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l71)

File: services/conversation/v1-exp.js (hunk 4 above)
Comment (chughts) [inline services/conversation/v1-exp.js:49]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l49)

File: services/conversation/v1-exp.js (hunk 5 above)
Comment (chughts) [inline services/conversation/v1-exp.js:48]: Codacy Issue found: [Identifier 'workspace_id' is not in camel case.](https://www.codacy.com/app/BetaWorks-NodeRED-Watson/node-red-node-watson/file/2776352362/issues/source?bid=3330694&fileBranchId=3363974#l48)

File: services/conversation/v1-exp.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      node.error('Missing property: msg.payload', msg);
+      return false;
+    }
+    return true;
+  }
+
+  function verifyInputs(node, msg, config) {
+    // workspaceid can be either configured in the node, 
+    // or sent into msg.params.workspace_id
+    if (!config.workspaceid || '' !== config.workspaceid) {
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:45]: Codacy Issue found: [Expected literal to be on the right side of !==.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l45)

File: services/conversation/v1-exp.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  function verifyPayload(node, msg) {
+    if (!msg.payload) {
+      this.status({fill:'red', shape:'ring', text:'missing payload'}); 
+      node.error('Missing property: msg.payload', msg);
+      return false;
+    }
+    return true;
+  }
+
+  function verifyInputs(node, msg, config) {
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:42]: Codacy Issue found: [Function 'verifyInputs' has a complexity of 5.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l42)

File: services/conversation/v1-exp.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+      return true;
+    }
+    node.error('Missing workspace_id. check node documentation.',msg);
+    return false;
+  }
+
+  function verifyServiceCredentials(node, msg) {
+    // If it is present the newly provided user entered key 
+    // takes precedence over the existing one. 
+    //node.apikey = sAPIKey || node.credentials.apikey;
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:59]: Codacy Issue found: [Expected space or tab after "//" in comment.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l59)

File: services/conversation/v1-exp.js (hunk 1 above)
Comment (yrezgui) [inline services/conversation/v1-exp.js:101]: Codacy Issue found: [Infix operators must be spaced.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l101)

File: services/conversation/v1-exp.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    params.input={text:msg.payload};
+    // call POST /message through SDK
+    node.service.message(params, function(err, body) {
+      processResponse(err,body,node,msg);
+    });
+  }
+
+  // This is the Watson Visual Recognition V3 Node
+  function WatsonConversationV1ExpNode (config) {
+    var node = this, b = false;
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:111]: Codacy Issue found: [Expected blank line after variable declarations.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l111)

File: services/conversation/v1-exp.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    return true;
+  }
+
+  function processResponse(err, body, node, msg) {
+    if (err != null && body == null) {
+      node.status({fill:'red', shape:'ring', 
+        text:'call to watson conversation service failed'}); 
+      msg.result = {};
+      if (err.code == null) {
+        msg.result['error']=err;
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:83]: Codacy Issue found: [Infix operators must be spaced.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l83)

File: services/conversation/v1-exp.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    msg.result = body;
+    msg.payload = 'see msg.result';
+    node.send(msg); 
+    node.status({});
+  }
+
+  function execute(params, node, msg) {
+    node.status({fill:'blue', shape:'dot' , text:'Calling Conversation service ...'});
+    params.workspace_id=node.workspaceid;
+    params.input={text:msg.payload};
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:102]: Codacy Issue found: [Infix operators must be spaced.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l102)

File: services/conversation/v1-exp.js [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+      return true;
+    } else if (!msg.params || '' !== msg.params.workspace_id) {
+      node.workspaceid = msg.params.workspace_id;
+      return true;
+    }
+    node.error('Missing workspace_id. check node documentation.',msg);
+    return false;
+  }
+
+  function verifyServiceCredentials(node, msg) {
```
Comment (yrezgui) [inline services/conversation/v1-exp.js:56]: Codacy Issue found: [Function 'verifyServiceCredentials' has a complexity of 5.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l56)

File: services/conversation/v1-exp.js (hunk 5 above)
Comment (yrezgui) [inline services/conversation/v1-exp.js:48]: Codacy Issue found: [Expected literal to be on the right side of !==.](https://www.codacy.com/app/node-red-nodes-watson/node-red-node-watson/file/2777516067/issues/source?bid=3225115&fileBranchId=3364374#l48)

**Comment by chughts** [comment]
I have marked the camel case deviations for the API params, as used in this file, as passed in codacy. These should now not be flagged. 

As the webhook for CLA is broken I am closing this pull request, to allow @ylecleach to resubmit, and trigger a fresh CLA request.

**Pull request closed without merging**